src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java

```diff
@@ -0,0 +1,215 @@
+        // very small assembly region objects, then we can collect them on the driver (or repartition them)
+        // for redistribution across the cluster, at which points the reads can be filled in. (See next two steps.)
+        JavaRDD<ReadlessAssemblyRegion> readlessAssemblyRegions = contigToGroupedStates
+                .flatMap(getReadlessAssemblyRegionsFunction(header, assemblyRegionArgs, includeReadsWithDeletionsInIsActivePileups));
+
+        // 4. Pull the assembly region boundaries down to the driver, so we can fill in reads.
+        List<ShardBoundary> assemblyRegionBoundaries = readlessAssemblyRegions
+                .map((Function<ReadlessAssemblyRegion, ShardBoundary>) FindAssemblyRegionsSpark::toShardBoundary)
+                .collect();
+
+        // 5. Fill in the reads. Each shard is an assembly region, with its overlapping reads.
+        JavaRDD<Shard<GATKRead>> assemblyRegionShardedReads = SparkSharder.shard(ctx, reads, GATKRead.class, header.getSequenceDictionary(), assemblyRegionBoundaries, shardingArgs.readShardSize);
+
+        // 6. Convert shards to assembly regions.
+        JavaRDD<AssemblyRegion> assemblyRegions = assemblyRegionShardedReads.map((Function<Shard<GATKRead>, AssemblyRegion>) shard -> toAssemblyRegion(shard, header));
+
+        // 7. Add reference and feature context.
+        return assemblyRegions.mapPartitions(getAssemblyRegionWalkerContextFunction(referenceFileName, bFeatureManager));
+    }
+
+    private static FlatMapFunction<Iterator<Shard<GATKRead>>, ActivityProfileStateRange> getActivityProfileStatesFunction(
+            final String referenceFileName,
+            final Broadcast<FeatureManager> bFeatureManager,
+            final SAMFileHeader header,
+            final Broadcast<Supplier<AssemblyRegionEvaluator>> supplierBroadcast,
+            final AssemblyRegionArgumentCollection assemblyRegionArgs,
+            final boolean includeReadsWithDeletionsInIsActivePileups) {
+        return (FlatMapFunction<Iterator<Shard<GATKRead>>, ActivityProfileStateRange>) shardedReadIterator -> {
+            ReferenceDataSource reference = referenceFileName == null ? null : new ReferenceFileSource(IOUtils.getPath(SparkFiles.get(referenceFileName)));
+            final FeatureManager features = bFeatureManager == null ? null : bFeatureManager.getValue();
+            AssemblyRegionEvaluator assemblyRegionEvaluator = supplierBroadcast.getValue().get(); // one AssemblyRegionEvaluator instance per Spark partition
+            final ReadsDownsampler readsDownsampler = assemblyRegionArgs.maxReadsPerAlignmentStart > 0 ?
+                    new PositionalDownsampler(assemblyRegionArgs.maxReadsPerAlignmentStart, header) : null;
+
+            return Utils.stream(shardedReadIterator)
+                    .map(shardedRead -> new ShardToMultiIntervalShardAdapter<>(shardedRead))
+                    // TODO: reinstate downsampling (not yet working)
```

Comment:
We should think about how we want to do this. Notably however, as of this implementation you need to add the same downsampling to the reads at step 5, as this is currently only hooked up to downsample for the band pass filtering but not on the reads when they get re-added to the filter. An obvious solution might involve persisiting the underlying bam but obviously thats expensive, the other is that we make the random seed in some way caclulatable, like base it on a hash of the read if necessary, that way we could get an exact mach with the regular haplotype caller.

Reply:
The reason that I excluded this is because I found an underlying bug in the downsampling code where it would not do the downsampling correctly unless the iterator was created twice. (This needs further investigation.)

But you are right that the reads downsampling needs to have been carried out for the reads too, which isn't currently happening in this code. Let's explore your ideas further.